Kraitchik.java

```diff
@@ -0,0 +1,320 @@
+	 * @param a historique de la ligne que l'on additionne à la seconde
+	 * @param b historique de la ligne à laquelle on additionne la première
+	 * @return nouvel historique d'opération de la seconde ligne
+	 */
+	public static ArrayList<Integer> majHistorique(ArrayList<Integer> a, ArrayList<Integer> b){
+		ArrayList<Integer> resultat = new ArrayList<>();
+		for(int i = 0; i < a.size(); i++){
+			if(!b.contains(a.get(i))){
+				resultat.add(a.get(i));
+			}
+		}
+		for(int i = 0; i < b.size(); i++){
+			if(!a.contains(b.get(i))){
+				resultat.add(b.get(i));
+			}
+		}
+		return resultat;
+	}
+
+	/**
+	 * Détermine un vecteur du noyau de la matrice transposée de @param M selon la méthode du pivot de Gauss.
+	 * La matrice n'est pas à proprement parler échelonnée car on effectue pas d'opération d'échange de ligne
+	 * @param M matrice à échelonner
+	 * @param historique historique d'opérations initial des lignes de la matrice
+	 * @return renvoie la liste des opérations effectuées pour obtenir un vecteur du noyau
+	 */
+	public static ArrayList<Integer> pivotGauss(ArrayList<ArrayList<Integer>> M, ArrayList<ArrayList<Integer>> historique){
+		int colonePivot = 0;
+		ArrayList<Integer> lignesPivot = new ArrayList<>();
+		while(!contientLigneNulle(M)){
+			for(int i = 0; i < M.size(); i++){
+				if(M.get(i).get(colonePivot) == 1 && !lignesPivot.contains(i)){
+					for(int j = i + 1; j < M.size(); j++){
+						if(M.get(j).get(colonePivot) == 1){
+							M.set(j,produitF2(M.get(i), M.get(j)));
+							historique.set(j, majHistorique(historique.get(i), historique.get(j)));
+						}
+					}
+					lignesPivot.add(i);
+				}
+			}
+			colonePivot++;
+		}
+		
+		for(int i = 0; i < M.size(); i++){
+			if(ligneNulle(M.get(i))){
+				return historique.get(i);
+			}
+		}
+		
+		return null;
+	}
+
+	/**
+	 * affiche une matrice
+	 * @param M matrice à afficher
+	 */
+	public static void printMatrice(ArrayList<ArrayList<Integer>> M){
+		for(int i = 0; i < M.size(); i++){
+			for(int j = 0; j < M.get(i).size(); j++){
+				System.out.print(M.get(i).get(j) + " ");
+			}
+			System.out.println();
+		}
+	}
+
+	/**
+	 * Factorise un entier @param n selon la méthode de Kraitchik
+	 * (Disclaimer) du fait de la représentation des entiers en Java, 
+	 * il n'est pas possible de factoriser de trop gros entiers
+	 * @param n entier à factoriser
+	 * @param b borne avec laquelle on teste la friabilité des résultats du polynome de Kraitchik
+	 * @param piB nombre de nombres premiers inférieurs ou égaux à @param b
+	 * @return renvoie deux facteurs de @param n
+	 */
+	public static ArrayList<Integer> factoKraitchik(int n, int b, int piB){
+		ArrayList<Integer> premiers = premiersInfB(b);
+		ArrayList<Integer> xi = new ArrayList<>();
+		ArrayList<Integer> qi = new ArrayList<>();
+		int x = Math.round(Math.round(Math.sqrt(n))) + 1;
+
+		while(xi.size() != piB + 1){
+			if(friable(polyKraitchik(x, n), b)){
```

Comment:
Pour n'avoir à le calculer qu'une fois
```suggestion
                        int q = polyKraitchik(x, n);
			if(friable(q, b)){
			    xi.add(x);
			    qi.add(q);
```